### PR TITLE
authWebview: Identity Center for CW

### PR DESCRIPTION
**Pull request by nkomonen-amazon**
Adds in Identity Center SSO sign in for CW in the auth webview
- Validates start url field
- Can sign out and will fully delete the connections

### Signing in to Identity Center
Kapture 2023-05-18 at 13 31 10

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/credentials/sso/shared.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*!
```
Comment (justinmk3) [inline src/credentials/sso/shared.ts:1]: can this live in sso/model.ts or somewhere else? we have 5+ `sso/*` modules, that is a lot for this one topic

File: src/credentials/vue/authForms/shared.vue [hunk 2]
```diff
@@ -10,6 +11,7 @@ const authFormsState = {
     [authForms.CREDENTIALS]: new CredentialsState(),
     [authForms.BUILDER_ID_CODE_WHISPERER]: new CodeWhispererBuilderIdState(),
     [authForms.BUILDER_ID_CODE_CATALYST]: new CodeCatalystBuilderIdState(),
+    [authForms.IDENTITY_CENTER_CODE_WHISPERER]: new CodeWhispererIdentityCenterState(),
```
Comment (justinmk3) [inline src/credentials/vue/authForms/shared.vue:14]:
```suggestion
    [authForms.IDENTITY_CENTER_CODEWHISPERER]: new CodeWhispererIdentityCenterState(),
```

File: src/credentials/vue/authForms/types.vue [hunk 3]
```diff
@@ -3,6 +3,7 @@ export const authForms = {
     CREDENTIALS: 'CREDENTIALS',
     BUILDER_ID_CODE_WHISPERER: 'BUILDER_ID_CODE_WHISPERER',
     BUILDER_ID_CODE_CATALYST: 'BUILDER_ID_CODE_CATALYST',
+    IDENTITY_CENTER_CODE_WHISPERER: 'IDENTITY_CENTER_CODE_WHISPERER',
```
Comment (justinmk3) [inline src/credentials/vue/authForms/types.vue:6]:
```suggestion
    IDENTITY_CENTER_CODEWHISPERER: 'IDENTITY_CENTER_CODEWHISPERER',
```

File: src/credentials/vue/serviceItemContent/codeWhispererContent.vue [hunk 4]
```diff
@@ -1,25 +1,33 @@
...
 import { defineComponent } from 'vue'
 import BuilderIdForm, { CodeWhispererBuilderIdState } from '../authForms/manageBuilderId.vue'
+import IndentityCenterForm, { CodeWhispererIdentityCenterState } from '../authForms/manageIdentityCenter.vue'
 import BaseServiceItemContent from './baseServiceItemContent.vue'
 import authFormsState, { AuthStatus } from '../authForms/shared.vue'
 
 export default defineComponent({
     name: 'CodeWhispererContent',
-    components: { BuilderIdForm },
+    components: { BuilderIdForm, IndentityCenterForm },
```
Comment (justinmk3) [inline src/credentials/vue/serviceItemContent/codeWhispererContent.vue:22]:
```suggestion
    components: { BuilderIdForm, IdentityCenterForm },
```

File: src/auth/ui/vue/show.ts [hunk 5]
```diff
@@ -136,6 +139,59 @@ export class AuthWebview extends VueWebview {
...
+        try {
+            await CodeWhispererAuth.instance.connectToEnterpriseSso(startUrl, regionId)
+        } catch (e) {
+            // This scenario will most likely be due to failing to connect from user error.
+            // When the sso login process fails (eg: wrong url) they will come back
+            // to the IDE and cancel the 'waiting for browser response'
+            if (CancellationError.isUserCancelled(e)) {
+                return
+            }
+        }
```
Comment (JadenSimon) [inline src/auth/ui/vue/show.ts:156]: Seems like this error should be bubbled up and shown in the webview if they didn't cancel?

File: src/auth/ui/vue/authForms/manageIdentityCenter.vue [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    protected override async _startIdentityCenterSetup(): Promise<void> {
+        const data = await this.getSubmittableDataOrThrow()
+        return client.startIdentityCenterSetup(data.startUrl, data.region)
+    }
+
+    private async getSubmittableDataOrThrow(): Promise<IdentityCenterData> {
+        return this._data as IdentityCenterData
+    }
```
Comment (JadenSimon) [inline src/auth/ui/vue/authForms/manageIdentityCenter.vue:218]: Feels clunky. The only difference between the concrete class and the abstract class is `isCodeWhispererIdentityCenterConnected` + an ID? Why not pass the id to the extension host and call the correct function there?

File: src/auth/sso/validation.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import { AuthType } from '../auth'
+import { isSsoConnection, SsoConnection, hasScopes } from '../connection'
+
+export function validateSsoUrl(auth: AuthType, url: string, requiredScopes?: string[]) {
+    const urlFormatError = validateSsoUrlFormat(url)
+    if (urlFormatError) {
+        return urlFormatError
+    }
+
+    return validateIsNewSsoUrlAsync(auth, url, requiredScopes)
```
Comment (justinmk3) [inline src/auth/sso/validation.ts:16]: unnecessary indirection. can use `.then(...)` to chain the result of `listConnections()` here

File: src/auth/sso/validation.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as vscode from 'vscode'
+import { UnknownError } from '../../shared/errors'
+import { AuthType } from '../auth'
+import { isSsoConnection, SsoConnection, hasScopes } from '../connection'
+
+export function validateSsoUrl(auth: AuthType, url: string, requiredScopes?: string[]) {
+    const urlFormatError = validateSsoUrlFormat(url)
```
Comment (justinmk3) [inline src/auth/sso/validation.ts:11]: unnecessary indirection. we're already in a function named `validateSsoUrl`. Too much granularity

**Pull request merged**